ARROW-13064: [C++] Implement select ('case when') function for fixed-width types

lidavidm opened the pull request:
This doesn't support variable-width types (e.g. strings) as the implementation here is columnwise. I will work on those separately (they require a rowwise implementation).

Also fixes a small bug in the CommonNumericType implementation (I noticed uint8 was getting promoted to int8).

lidavidm commented:
Note there's some code here for fixed-width types that duplicates what's in ARROW-9430. They should probably get unified.

jorisvandenbossche commented:
To start my usual round of name bike-shedding ;), I would propose to not call this "select", to avoid the confusion with SQL's `SELECT` / dplyr's `select()` function (I think the "select" name only comes from `np.select`?) 
Given this is for SQL's `CASE`, maybe something like "select_case" or "case_when"?

lidavidm commented:
`case_when` sounds good to me. Or perhaps `switch_case` by analogy to `if_else`.

pitrou commented:
Ok for `case_when`, another possibility is `cond` (like in Lisp ;-)).

lidavidm commented:
Renamed to `case_when`.

pitrou commented:
It looks like this needs rebasing.

lidavidm commented:
Thanks for the heads up, rebased.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+                        const int64_t offset, const int64_t length) {
+    CType* out_values = reinterpret_cast<CType*>(raw_out_values);
+    const CType* in_values = array.GetValues<CType>(1);
+    std::copy(in_values + offset, in_values + offset + length, out_values + offset);
```
At some point, we should perhaps try to find out whether `std::copy` is as performant as `memcpy`.
cc @bkietz for opinions.

bkietz replied:
std::copy allows the ranges to overlap (unlike memcpy), so for simple pointers like this it gets [inlined to memmove](https://godbolt.org/z/K1ov93rYb). Memmove can be slower (or faster) than memcpy; depends on your libc. It shouldn't differ by a very wide margin though so I don't think it's necessary to avoid std::copy

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+    uint8_t* next = out_values + (width * offset);
+    const auto& scalar = checked_cast<const FixedSizeBinaryScalar&>(values);
+    // Scalar may have null value buffer
+    if (!scalar.value) return;
```
If the scalar isn't valid, we should zero-initialize the destination memory area.

lidavidm replied:
It looks like this is because the various binary scalars have constructors that take only a DataType - for the non-fixed-size scalars, we can construct an empty buffer, but for the fixed-size scalar, doing this would require allocation. And just removing that constructor for FixedSizeBinaryScalar just pushes the problem around (e.g. now MakeNullScalar needs to do allocation, and it complicates the overloading it does).

lidavidm replied:
At least this check should be `if (!scalar.is_valid) return` though, so we'll trip if we somehow have a non-null scalar with a null value buffer.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+        cond->type = boolean();
+      }
+      if (cond->type->id() != Type::BOOL) {
+        return Status::Invalid("Condition arguments must be boolean, but argument ", i,
```
`TypeError`?

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+    }
+    EnsureDictionaryDecoded(&value_types);
+    if (auto type = CommonNumeric(value_types)) {
+      ReplaceTypes(type, &value_types);
```
This doesn't seem to do anything special apart from duplicating `type` accross `value_types`. Instead you can just reuse `type` below (call it `promoted_value_type`?).

lidavidm replied:
AIUI this is needed because then the arguments will get implicitly cast afterwards.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+    const DataType& common_values_type = *value_types.front().type;
+    auto next_type = value_types.cbegin();
+    for (size_t i = 0; i < values->size(); i += 2) {
+      if (!common_values_type.Equals(next_type->type)) {
```
Given the definition of `ReplaceTypes`, this check doesn't seem necessary?
cc @bkietz  for a second opinion.

bkietz replied:
It seems to me we could restructure "case_when" and yield a less daunting interface:

```python
case(
  when(cond_0, value_0),
  when(cond_1, value_1),
  value_else
)
```

Where `when` masks slots with null wherever its condition is not true and `case` is only a variadic coalescing function (takes the first non null).

We'd be allocating a new null bitmap on each call to `when` which is not ideal. However typing is far clearer (`when(cond: Boolean, value: T): T`, `case(...values: T): T`) and case/when can be independently unit tested.

@pitrou  what do you think?

bkietz replied:
Note: variadic coalesce is independently useful https://issues.apache.org/jira/browse/ARROW-13136

bkietz replied:
Another way to transpose this would be:

```python
case(
  first_true_in(cond_0, cond1),
  value0,
  value1,
  value_else
)
```

Where `first_true_in` returns the index of the first `true` and `case`'s first argument is an index into the rest of its arguments:

```
assert first_true_in([True, False, False], [True, True, False]) == [0, 1, None]
assert case([0, 1, None], 'a b c'.split(), 'd e f'.split(), 'x y z'.split()) == 'a e z'.split()
```

lidavidm replied:
In the second example, case would essentially be 'choose' (ARROW-13220). I'm curious which would be faster; the coalesce PR is implemented columnwise (for fixed-width types) and tries to do block copies, while the 'choose' PR I'm looking at is necessarily rowwise. Probably depends on the distribution of the inputs.

Might the masking 'when' kernel and 'first_true_in' be useful kernels to have anyways?

bkietz replied:
I think the masking 'when' and 'first_true_in' are definitely independently useful

lidavidm replied:
I could get choose/coalesce in first, then put up candidate PRs for case-when to compare the implementations.

bkietz replied:
SGTM, thanks!

jorisvandenbossche replied:
I think that 

```
case(
  when(cond_0, value_0),
  when(cond_1, value_1),
  value_else
)
```

is not necessarily identical to `case_when(cond_0, value_0, cond_1, value_1, value_else)` as done in this PR, when it comes to nulls in the values? 

For example, if you have a null in `value_0`, it should be chosen if the corresponding bit of `cond_0` is True. However, if you implement that with a masking `when` and then coalesce (taking the first non null), it would not pick the null value.

lidavidm replied:
Hmm, you're right…

We can still try the 'choose' based version though I would expect that to be slower since it works rowwise.

lidavidm replied:
I put up PRs for both choose: #10642 and coalesce: #10608 (though as you point out, we can't quite use coalesce for this).

lidavidm replied:
A quick check with choose shows that it's only about 2/3 as fast:

```
CaseWhenBench32/1048576/0              48183534 ns     48183323 ns           13 bytes_per_second=334.659M/s
CaseWhenBench64/1048576/0              48610718 ns     48610432 ns           13 bytes_per_second=660.866M/s
CaseWhenBench32/1048576/99             48226819 ns     48226677 ns           15 bytes_per_second=334.327M/s
CaseWhenBench64/1048576/99             48891957 ns     48892172 ns           13 bytes_per_second=656.996M/s
```

though note we have to call both first_true_in, then fill_null, to get the 'else' behavior which doesn't help.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else_test.cc:
```diff
@@ -316,5 +318,165 @@ TEST_F(TestIfElseKernel, IfElseDispatchBest) {
+                Check({boolean(), utf8(), int32()}, {boolean(), int32(), int32()}));
+  ASSERT_RAISES(Invalid,
+                Check({int32(), int32(), int32()}, {boolean(), int32(), int32()}));
+}
```
Somewhere, you should also test what happens when `case_when` is called with zero arguments.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+};
+template <typename Type>
+struct CopyFixedWidth<Type, enable_if_same<Type, FixedSizeBinaryType>> {
+  static void CopyScalar(const Scalar& values, uint8_t* out_values, const int64_t offset,
```
Instead of `uint8_t* out_values`, you may want this to take a `ArrayData* out`, since you'll need it for non-fixed-width types?

lidavidm replied:
For non-fixed-width types, it might be simpler to handle it with a builder and AppendScalar?

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+  // Initialize values buffer
+  uint8_t* out_values = output->buffers[1]->mutable_data();
+  if (have_else_arg) {
+    // Copy 'else' value into output
```
This seems a bit excessive?

lidavidm replied:
This seems to be faster as written, oddly.

Before:

```
-----------------------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------
CaseWhenBench32/1048576/0              31933112 ns     31932368 ns           22 bytes_per_second=504.974M/s
CaseWhenBench64/1048576/0              33170481 ns     33168735 ns           21 bytes_per_second=968.533M/s
CaseWhenBench32/1048576/99             32487300 ns     32487411 ns           21 bytes_per_second=496.299M/s
CaseWhenBench64/1048576/99             33682029 ns     33680901 ns           21 bytes_per_second=953.715M/s
CaseWhenBench32Contiguous/1048576/0     7255445 ns      7255387 ns           96 bytes_per_second=1.632G/s
CaseWhenBench64Contiguous/1048576/0     7932437 ns      7932171 ns           88 bytes_per_second=2.97013G/s
CaseWhenBench32Contiguous/1048576/99    7526742 ns      7526709 ns           92 bytes_per_second=1.57303G/s
CaseWhenBench64Contiguous/1048576/99    8172498 ns      8172239 ns           83 bytes_per_second=2.88261G/s
```

After:

```
-----------------------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------
CaseWhenBench32/1048576/0              44166172 ns     44165634 ns           16 bytes_per_second=365.103M/s
CaseWhenBench64/1048576/0              44605356 ns     44603995 ns           16 bytes_per_second=720.227M/s
CaseWhenBench32/1048576/99             44867670 ns     44867051 ns           16 bytes_per_second=359.361M/s
CaseWhenBench64/1048576/99             45077818 ns     45076721 ns           15 bytes_per_second=712.607M/s
CaseWhenBench32Contiguous/1048576/0    17757494 ns     17757271 ns           39 bytes_per_second=682.819M/s
CaseWhenBench64Contiguous/1048576/0    18236327 ns     18235892 ns           38 bytes_per_second=1.29193G/s
CaseWhenBench32Contiguous/1048576/99   15051281 ns     15051008 ns           39 bytes_per_second=805.518M/s
CaseWhenBench64Contiguous/1048576/99   15504081 ns     15503998 ns           45 bytes_per_second=1.51944G/s
```

pitrou commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+    const uint8_t* cond_values = cond_array.buffers[1]->data();
+    int64_t offset = 0;
+    // If no valid buffer, visit mask & value bitmap simultaneously
+    if (!cond_array.MayHaveNulls()) {
```
You can use `GetNullCount()` I think?

pitrou reviewed (changes requested):
Very nice, thank you. Do you want to add a couple benchmarks for this?

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_if_else_test.cc:
```diff
@@ -316,5 +318,165 @@ TEST_F(TestIfElseKernel, IfElseDispatchBest) {
   CheckDispatchBest(name, {null(), uint8(), int8()}, {boolean(), int16(), int16()});
 }
 
+void CheckVarArgs(const std::string& name, const std::vector<Datum>& inputs,
```
It's also somewhat overlapping with CheckScalar now that I look at it. I'll take a look and see if I can't consolidate all three. (I may do so in a different PR until we decide what to do with the implementation here, if we want to split it up into separate 'case' and 'when' functions or not as suggested.)

lidavidm commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -589,6 +589,21 @@ ARROW_EXPORT
 
+/// \brief CaseWhen behaves like a switch/case or if-else if-else statement: for
+/// each row, select the first value for which the corresponding condition is
+/// true, or (if given) select the 'else' value, else emit null.
```
Yes, null is the same as false.

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_if_else_test.cc:
```diff
@@ -316,5 +318,165 @@ TEST_F(TestIfElseKernel, IfElseDispatchBest) {
+
+TYPED_TEST(TestCaseWhenNumeric, FixedSize) {
+  auto type = default_type_instance<TypeParam>();
+  CheckCaseWhenCases(type, "10", "42");
```
I'll inline all tests as I've done with coalesce and choose.

lidavidm commented:
I've changed this so the signature is `CaseWhen<T>(struct(bool...), T...)` instead; if this looks more reasonable, we can proceed, else I'll revert it.

jorisvandenbossche commented:
What's the advantage of using a struct of boolean arrays as first argument instead of a vector? (eg for a struct you need to give it (dummy) names)

lidavidm commented:
@bkietz and I discussed this. Mostly it simplifies the function signature a lot and gets rid of the custom typechecking code in DispatchBest. It does require materializing scalar conditions and would need callers to do some pre-processing if they wish to avoid that (e.g. dropping any values associated with a scalar false/null condition, using values associated with a scalar true condition as the 'else' clause).

It's mostly a point of discussion. I realized I forgot to benchmark so let me do that now.

lidavidm commented:
This is quite faster though (actually almost suspiciously so; also, the benchmark loop doesn't include constructing a structarray from individual boolean arrays).

Struct:
```
CaseWhenBench32/1048576/0               5780694 ns      5780134 ns          116 bytes_per_second=2.72434G/s
CaseWhenBench64/1048576/0               6079444 ns      6078645 ns          115 bytes_per_second=5.16103G/s
CaseWhenBench32/1048576/99              5717893 ns      5717306 ns          122 bytes_per_second=2.75402G/s
CaseWhenBench64/1048576/99              6160254 ns      6159488 ns          110 bytes_per_second=5.09281G/s
CaseWhenBench32Contiguous/1048576/0     1652127 ns      1645047 ns          423 bytes_per_second=7.19786G/s
CaseWhenBench64Contiguous/1048576/0     1986278 ns      1986066 ns          334 bytes_per_second=11.8624G/s
CaseWhenBench32Contiguous/1048576/99    1826875 ns      1826675 ns          377 bytes_per_second=6.48156G/s
CaseWhenBench64Contiguous/1048576/99    2415119 ns      2414697 ns          312 bytes_per_second=9.75582G/s
```

Alternating arguments:
```
CaseWhenBench32/1048576/0              44467484 ns     44467216 ns           16 bytes_per_second=362.627M/s
CaseWhenBench64/1048576/0              44722665 ns     44721786 ns           16 bytes_per_second=718.33M/s
CaseWhenBench32/1048576/99             45402157 ns     45399285 ns           16 bytes_per_second=355.148M/s
CaseWhenBench64/1048576/99             45465935 ns     45464130 ns           15 bytes_per_second=706.534M/s
CaseWhenBench32Contiguous/1048576/0    17973260 ns     17972895 ns           39 bytes_per_second=674.627M/s
CaseWhenBench64Contiguous/1048576/0    18549381 ns     18548768 ns           37 bytes_per_second=1.27014G/s
CaseWhenBench32Contiguous/1048576/99   18203192 ns     18202563 ns           39 bytes_per_second=666.052M/s
CaseWhenBench64Contiguous/1048576/99   18817987 ns     18817296 ns           37 bytes_per_second=1.2519G/s
```

lidavidm commented:
Any other thoughts here? If we're ok with CaseWhen(struct(bool...), T...) over CaseWhen(bool, T, bool, T, ...) I can make this kernel also support writing into slices.

bkietz commented:
I'm fine with `CaseWhen(struct(bool...), T...)`, and I'd love for this to support writing into slices. Could you also file a follow up JIRA for simplification of call expressions involving case_when?

lidavidm commented:
Filed ARROW-13325.

lidavidm commented:
Updated to support can_write_into_slices. Also, updated the benchmark bytes_per_second scaling to reflect the size of the output and not the size of all the inputs so the numbers are more reasonable.

```
-----------------------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------
CaseWhenBench32/1048576/0               6065579 ns      6065602 ns           97 bytes_per_second=659.456M/s
CaseWhenBench64/1048576/0               6485780 ns      6485814 ns           93 bytes_per_second=1.20455G/s
CaseWhenBench32/1048576/99              6280547 ns      6280511 ns          114 bytes_per_second=636.831M/s
CaseWhenBench64/1048576/99              6694807 ns      6694710 ns          105 bytes_per_second=1.16686G/s
CaseWhenBench32Contiguous/1048576/0     1761802 ns      1761781 ns          378 bytes_per_second=2.21722G/s
CaseWhenBench64Contiguous/1048576/0     2267441 ns      2267379 ns          280 bytes_per_second=3.44561G/s
CaseWhenBench32Contiguous/1048576/99    1897469 ns      1897449 ns          325 bytes_per_second=2.05849G/s
CaseWhenBench64Contiguous/1048576/99    2515244 ns      2515231 ns          275 bytes_per_second=3.10578G/s
```

bkietz commented on docs/source/cpp/compute.rst:
```diff
@@ -859,50 +859,60 @@ Structural transforms
++--------------------------+------------+------------------------------------------------+---------------------+---------+
+
+* \(1) This function acts like a SQL 'case when' statement or switch-case. The
+  input is any number of alternating Boolean and value data, followed by an
```
This (and some of the other comments) need to be updated with the new signature

bkietz commented on cpp/src/arrow/compute/kernels/scalar_if_else_benchmark.cc:
```diff
@@ -97,6 +97,99 @@ static void IfElseBench32Contiguous(benchmark::State& state) {
+
+  ASSERT_OK_AND_ASSIGN(auto trues, MakeArrayFromScalar(BooleanScalar(true), len / 3));
+  ASSERT_OK_AND_ASSIGN(auto falses, MakeArrayFromScalar(BooleanScalar(false), len / 3));
+  auto null_scalar = MakeNullScalar(boolean());
+  ASSERT_OK_AND_ASSIGN(auto nulls,
+                       MakeArrayFromScalar(*null_scalar, len - 2 * (len / 3)));
```
```suggestion
  ASSERT_OK_AND_ASSIGN(auto nulls, MakeArrayOfNull(boolean(), len - 2 * (len / 3)));
```

bkietz commented on cpp/src/arrow/compute/kernels/scalar_if_else_benchmark.cc:
```diff
@@ -97,6 +97,99 @@ static void IfElseBench32Contiguous(benchmark::State& state) {
+
+  random::RandomArrayGenerator rand(/*seed=*/0);
+
+  auto cond1 = std::static_pointer_cast<BooleanArray>(
+      rand.ArrayOf(boolean(), len, /*null_probability=*/0.01));
+  auto cond2 = std::static_pointer_cast<BooleanArray>(
+      rand.ArrayOf(boolean(), len, /*null_probability=*/0.01));
+  auto cond3 = std::static_pointer_cast<BooleanArray>(
+      rand.ArrayOf(boolean(), len, /*null_probability=*/0.01));
```
```suggestion
  auto fld = field("cond", boolean(), key_value_metadata({{"null_probability", "1.0"}}));
  auto cond = rand.ArrayOf(field("", struct_({fld, fld, fld}), len);
```
?

bkietz reviewed (changes requested):
Some minor comments, otherwise this looks good to go

lidavidm commented:
Fixed docstrings and fixed a TODO I noticed about fully initializing the output buffer.

Also added a benchmark case for when both the cond struct array and the child cond boolean arrays can have nulls. This case is especially terrible, I made a slight optimization to eliminate one of the more egregious offenders I saw in perf, but it's still bad even then:

```
-----------------------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------
CaseWhenBench64/1048576/0               6045379 ns      6045204 ns          125 bytes_per_second=1.29235G/s
CaseWhenBench64/1048576/99              6106510 ns      6106346 ns          123 bytes_per_second=1.27929G/s
CaseWhenBench64OuterNulls/1048576/0    32735686 ns     32733991 ns           23 bytes_per_second=244.394M/s
CaseWhenBench64OuterNulls/1048576/99   34098317 ns     34097473 ns           21 bytes_per_second=234.599M/s
CaseWhenBench64Contiguous/1048576/0     2229728 ns      2229687 ns          338 bytes_per_second=3.50386G/s
CaseWhenBench64Contiguous/1048576/99    2456084 ns      2456085 ns          237 bytes_per_second=3.18058G/s
```

bkietz commented:
@lidavidm what would you think about just raising an error for top level nulls? It doesn't seem like a useful case to me

lidavidm commented:
Should be fine. It would certainly trim down the inner loop a lot.

lidavidm commented:
Removed support for toplevel nulls.

lidavidm commented:
@bkietz anything left either here or in #10608?

jorisvandenbossche commented:
> Any other thoughts here? If we're ok with CaseWhen(struct(bool...), T...) over CaseWhen(bool, T, bool, T, ...)

As mentioned on the meeting, I don't have a strong opinion on the exact signature (assuming you can create the inputs easily in the bindings either way), and if it simplifies the implementation / signature quite a bit, that sounds as a good reason.

I quickly tried it out in Python:

```
In [7]: cond = pc.project(pa.array([True, False, None]), pa.array([False, True, None]), field_names=[b"a", b"b"])

In [8]: pc.case_when(cond, pa.array([1, 2, 3]), pa.array([11, 12, 13]))
Out[8]: 
<pyarrow.lib.Int64Array object at 0x7f104124c820>
[
  1,
  12,
  null
]
```

What I find a little bit annoying is that I have to provide "dummy" names (that are never used) to my boolean conditions in `project` (maybe we could have some default for this so for this specific case specifying names is not required?) 
(I also noticed that the field names needed to be bytes and not strings, I suppose that's an error in the cython bindings, will open an issue about that).  
One other issue is that this uses the "project" kernel, which I thought was not really meant for users? (ARROW-11206) I could of course also have used `pa.StructArray.from_arrays` I assume.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_if_else.cc:
```diff
@@ -676,7 +677,353 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+Status ExecArrayCaseWhen(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  const auto& conds_array = *batch.values[0].array();
+  if (conds_array.GetNullCount() > 0) {
+    return Status::Invalid("cond struct must not have nulls");
```
```suggestion
    return Status::Invalid("cond struct must not have top-level nulls");
```

? (not fully sure if this is the correct terminology, but just to clarify about which nulls this is, as each individual condition field can still have nulls)

bkietz reviewed (approved):
LGTM.

@jorisvandenbossche I'll make a PR for ARROW-11206 shortly and allow `make_struct` to be called with no names provided (defaulting to empty names or `str(field_index)` or so). That should enable you to write

```c++
pc.case_when(
    pc.make_struct([True, False, None],
                   [False, True, None]),
    [1, 2, 3],
    [11, 12, 13])
```

jorisvandenbossche commented:
@bkietz sounds good!
